File: VB/DXGridSample/GridSelectingBehavior.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports DevExpress.Xpf.Grid
Imports System.Windows.Input
Imports System.Windows
Imports System.Windows.Media
Imports DevExpress.Xpf.Core.Native
Imports System.Windows.Controls.Primitives
Imports System.ComponentModel
Imports System.Windows.Controls
Imports DevExpress.Mvvm.UI.Interactivity

Namespace DXGrid_AssignComboBoxToColumn
	Public Class SelectionBehaviorSettings
		Public DragArea As Integer = 4
		Public StartScrollArea As Integer = 25
		Public UnselectedRowCountWhileScrolling As Integer = 0

		Public Function IsDragging(ByVal pt1 As Point, ByVal pt2 As Point) As Boolean
			Return Math.Abs(pt1.X - pt2.X) > DragArea OrElse Math.Abs(pt1.Y - pt2.Y) > DragArea
		End Function
		Public Function IsMouseOverTopScrollingArea(ByVal pt As Point) As Boolean
			Return pt.Y < StartScrollArea
		End Function
		Public Function IsMouseOverBottomScrollingArea(ByVal pt As Point, ByVal maxHeight As Double) As Boolean
			Return pt.Y > maxHeight - StartScrollArea
		End Function
		Public Function IsMouseOverLeftScrollingArea(ByVal pt As Point) As Boolean
			Return pt.X < StartScrollArea
		End Function
		Public Function IsMouseOverRightScrollingArea(ByVal pt As Point, ByVal maxWidth As Double) As Boolean
			Return pt.X > maxWidth - StartScrollArea
		End Function
	End Class
	Public Class SelectionInfo
		Private privateMousePoint As Point
		Public Property MousePoint() As Point
			Get
				Return privateMousePoint
			End Get
			Set(ByVal value As Point)
				privateMousePoint = value
			End Set
		End Property
		Private privateRowHandle As Integer
		Public Property RowHandle() As Integer
			Get
				Return privateRowHandle
			End Get
			Set(ByVal value As Integer)
				privateRowHandle = value
			End Set
		End Property
		Private privateColumn As GridColumn
		Public Property Column() As GridColumn
			Get
				Return privateColumn
			End Get
			Set(ByVal value As GridColumn)
				privateColumn = value
			End Set
		End Property
		Public Overridable Sub Clear()
			MousePoint = InvalidPoint
			RowHandle = GridControl.InvalidRowHandle
			Column = Nothing
		End Sub
		Public Overridable Function IsEmptyInfo() As Boolean
			Return MousePoint = InvalidPoint
		End Function

		Private Shared ReadOnly InvalidPoint As New Point(-10000, 10000)
	End Class
	Public Class StartSelectionInfo
		Inherits SelectionInfo
		Private privateIsSelectionStarted As Boolean
		Public Property IsSelectionStarted() As Boolean
			Get
				Return privateIsSelectionStarted
			End Get
			Set(ByVal value As Boolean)
				privateIsSelectionStarted = value
			End Set
		End Property
		Public ReadOnly Property IsLeftMouseButtonPressed() As Boolean
			Get
				Return Mouse.LeftButton = MouseButtonState.Pressed
			End Get
		End Property
		Public Overrides Sub Clear()
			MyBase.Clear()
			IsSelectionStarted = False
		End Sub
	End Class
	Public Class GridSelectingBehavior
		Inherits Behavior(Of GridControl)
		Private IsLogicBlocked As Boolean
		Private privateSettings As SelectionBehaviorSettings
		Public Property Settings() As SelectionBehaviorSettings
			Get
				Return privateSettings
			End Get
			Set(ByVal value As SelectionBehaviorSettings)
				privateSettings = value
			End Set
		End Property
		Public ReadOnly Property Grid() As GridControl
			Get
				Return AssociatedObject
			End Get
		End Property
		Public ReadOnly Property View() As TableView
			Get
				Return CType(Grid.View, TableView)
			End Get
		End Property
		Private privateScrollElement As IScrollInfo
		Protected Friend Property ScrollElement() As IScrollInfo
			Get
				Return privateScrollElement
			End Get
			Private Set(ByVal value As IScrollInfo)
				privateScrollElement = value
			End Set
		End Property
		Private privateDataArea As FrameworkElement
		Protected Friend Property DataArea() As FrameworkElement
			Get
				Return privateDataArea
			End Get
			Private Set(ByVal value As FrameworkElement)
				privateDataArea = value
			End Set
		End Property
		Private privateStartSelectionInfo As StartSelectionInfo
		Protected Property StartSelectionInfo() As StartSelectionInfo
			Get
				Return privateStartSelectionInfo
			End Get
			Private Set(ByVal value As StartSelectionInfo)
				privateStartSelectionInfo = value
			End Set
		End Property
		Private privateCurrentSelectionInfo As SelectionInfo
		Protected Property CurrentSelectionInfo() As SelectionInfo
			Get
				Return privateCurrentSelectionInfo
			End Get
			Private Set(ByVal value As SelectionInfo)
				privateCurrentSelectionInfo = value
			End Set
		End Property
		Private privateScrollController As ScrollController
		Protected Property ScrollController() As ScrollController
			Get
				Return privateScrollController
			End Get
			Set(ByVal value As ScrollController)
				privateScrollController = value
			End Set
		End Property

		Public Sub New()
			Settings = New SelectionBehaviorSettings()
			StartSelectionInfo = New StartSelectionInfo()
			CurrentSelectionInfo = New SelectionInfo()
			ScrollController = New ScrollController(Me)
			AddHandler ScrollController.ScrollDown, AddressOf OnScrollControllerScrollDown
			AddHandler ScrollController.ScrollUp, AddressOf OnScrollControllerScrollUp
			AddHandler ScrollController.ScrollLeft, AddressOf OnScrollControllerScrollLeft
			AddHandler ScrollController.ScrollRight, AddressOf OnScrollControllerScrollRight
		End Sub

		#Region " INITIALIZATION"
		Protected Overrides Sub OnAttached()
			MyBase.OnAttached()
			AddHandler Grid.Loaded, AddressOf OnGridLoaded
		End Sub
		Protected Overrides Sub OnDetaching()
			RemoveHandler Grid.PreviewMouseMove, AddressOf OnGridPreviewMouseMove
			RemoveHandler Grid.PreviewMouseDown, AddressOf OnGridPreviewMouseDown
			RemoveHandler Grid.PreviewMouseUp, AddressOf OnGridPreviewMouseUp
			RemoveHandler Grid.Loaded, AddressOf OnGridLoaded
			Dim dpd = DependencyPropertyDescriptor.FromProperty(GridControl.SelectionModeProperty, GetType(GridControl))
			dpd.AddValueChanged(Grid, AddressOf OnSelectionModeChanged)
			MyBase.OnDetaching()
		End Sub

		Private Sub OnSelectionModeChanged(ByVal sender As Object, ByVal e As EventArgs)
			If Grid.SelectionMode <> MultiSelectMode.Row OrElse Grid.SelectionMode <> MultiSelectMode.MultipleRow Then
				Grid.UnselectAll()
				IsLogicBlocked = True
			Else
				IsLogicBlocked = False
			End If
		End Sub
		Private Sub OnGridLoaded(ByVal sender As Object, ByVal e As RoutedEventArgs)
			AddHandler View.LayoutUpdated, AddressOf OnViewLayoutUpdated
		End Sub
		Private Sub OnViewLayoutUpdated(ByVal sender As Object, ByVal e As EventArgs)
			If Grid.SelectionMode = MultiSelectMode.None Then
				IsLogicBlocked = True
			End If
			DataArea = LayoutHelper.FindElement(View, AddressOf IsScrollContentPresenterAndHasRowPresenterGridAsParent)
			If DataArea Is Nothing Then
				Return
			End If
			ScrollElement = CType(LayoutHelper.FindElement(DataArea, AddressOf IsDataPresenter), DataPresenter)
			If ScrollElement Is Nothing Then
				Return
			End If

			RemoveHandler View.LayoutUpdated, AddressOf OnViewLayoutUpdated

			AddHandler Grid.PreviewMouseMove, AddressOf OnGridPreviewMouseMove
			AddHandler Grid.PreviewMouseDown, AddressOf OnGridPreviewMouseDown
			AddHandler Grid.PreviewMouseUp, AddressOf OnGridPreviewMouseUp
		End Sub

		Private Function IsScrollContentPresenter(ByVal e As FrameworkElement) As Boolean
			Return e.Name = "PART_ScrollContentPresenter"
		End Function
		Private Function IsRowPresenterGrid(ByVal e As DependencyObject) As Boolean
			Return (TypeOf e Is Grid) AndAlso ((CType(e, FrameworkElement)).Name = "rowPresenterGrid")
		End Function
		Private Function IsDataPresenter(ByVal e As FrameworkElement) As Boolean
			Return TypeOf e Is DataPresenter
		End Function
		Private Function IsScrollContentPresenterAndHasRowPresenterGridAsParent(ByVal e As FrameworkElement) As Boolean
			Return IsScrollContentPresenter(e) AndAlso LayoutHelper.FindLayoutOrVisualParentObject(e, AddressOf IsRowPresenterGrid) IsNot Nothing
		End Function
		#End Region

		Private Sub OnGridPreviewMouseUp(ByVal sender As Object, ByVal e As MouseButtonEventArgs)

			If (Not IsLogicBlocked) Then
				If (Not StartSelectionInfo.IsSelectionStarted) Then
					Return
				End If
				Mouse.Capture(Nothing)
				ScrollController.StopHorizontalScrolling()
				ScrollController.StopVerticalScrolling()
				StartSelectionInfo.Clear()
				CurrentSelectionInfo.Clear()
				e.Handled = True
			End If
		End Sub
		Private Sub OnGridPreviewMouseDown(ByVal sender As Object, ByVal e As MouseButtonEventArgs)
			If (Not IsLogicBlocked) Then
				StartSelectionInfo.Clear()
				Dim hitInfo As TableViewHitInfo = View.CalcHitInfo(TryCast(e.OriginalSource, DependencyObject))
				If (Not hitInfo.InRow) AndAlso (Not hitInfo.InRowCell) Then
					Return
				End If
				StartSelectionInfo.MousePoint = e.GetPosition(DataArea)
				StartSelectionInfo.RowHandle = hitInfo.RowHandle
				StartSelectionInfo.Column = hitInfo.Column
			End If
		End Sub
		Private Sub OnGridPreviewMouseMove(ByVal sender As Object, ByVal e As MouseEventArgs)
			If (Not IsLogicBlocked) Then
				If (Not StartSelectionInfo.IsLeftMouseButtonPressed) OrElse StartSelectionInfo.IsEmptyInfo() Then
					Return
				End If
				If (Not StartSelectionInfo.IsSelectionStarted) Then
					If (Not Settings.IsDragging(StartSelectionInfo.MousePoint, e.GetPosition(DataArea))) Then
						Return
					End If
					Mouse.Capture(DataArea, CaptureMode.SubTree)
					StartSelectionInfo.IsSelectionStarted = True
				End If

				UpdateCurrentSelectionInfo()

				UpdateVerticalScrolling(CurrentSelectionInfo.MousePoint)
				UpdateHorizontalScrolling(CurrentSelectionInfo.MousePoint)

				If (Not ScrollController.IsScrollWorking) Then
					UpdateSelection()
				End If

				ScrollController.UpdateVerticalScrollTimerInterval(DataArea.ActualHeight, e.GetPosition(DataArea).Y)
				ScrollController.UpdateHorizontalScrollTimerInterval(DataArea.ActualWidth, e.GetPosition(DataArea).X)
				e.Handled = True
			End If
		End Sub

		'TODO: improve this method
		Private Sub UpdateCurrentSelectionInfo()
			CurrentSelectionInfo.MousePoint = Mouse.GetPosition(DataArea)
			Dim pt As Point = CurrentSelectionInfo.MousePoint
			'Inside DataArea
			If pt.X > 0 AndAlso pt.Y > 0 AndAlso pt.X < DataArea.ActualWidth AndAlso pt.Y < DataArea.ActualHeight Then
				Dim result As HitTestResult = VisualTreeHelper.HitTest(DataArea, CurrentSelectionInfo.MousePoint)
				If result IsNot Nothing Then
					Dim hittedObject As DependencyObject = result.VisualHit
					Dim hitInfo As TableViewHitInfo = View.CalcHitInfo(hittedObject)
					If hitInfo.RowHandle <> GridControl.InvalidRowHandle Then
						CurrentSelectionInfo.RowHandle = hitInfo.RowHandle
					End If
					If hitInfo.Column IsNot Nothing Then
						CurrentSelectionInfo.Column = hitInfo.Column
					End If
				End If
				Return
			End If
			'Outside DataArea
			CurrentSelectionInfo.RowHandle = GetRowhandle(CurrentSelectionInfo.MousePoint.Y)
			'Select to right
			Dim rightColumn, rightColumnWhenUnselect As GridColumn
			GetRightVisibleColumn(rightColumn, rightColumnWhenUnselect)
			If pt.X >= DataArea.ActualWidth AndAlso CurrentSelectionInfo.Column IsNot Nothing AndAlso StartSelectionInfo.Column.VisibleIndex <= CurrentSelectionInfo.Column.VisibleIndex Then
				If rightColumn.VisibleIndex > CurrentSelectionInfo.Column.VisibleIndex Then
					CurrentSelectionInfo.Column = rightColumn
				End If
				Return
			End If
			'Unselecting to right
			If pt.X >= DataArea.ActualWidth AndAlso StartSelectionInfo.Column.VisibleIndex >= CurrentSelectionInfo.Column.VisibleIndex Then
				If rightColumnWhenUnselect.VisibleIndex >= CurrentSelectionInfo.Column.VisibleIndex Then
					CurrentSelectionInfo.Column = rightColumnWhenUnselect
				End If
				Return
			End If
			'Select to left
			If pt.X <= 0 Then
				Dim col As GridColumn = GetLeftVisibleColumn()
				CurrentSelectionInfo.Column = col
				Return
			End If
		End Sub
		Private Sub GetRightVisibleColumn(<System.Runtime.InteropServices.Out()> ByRef rightColumn As GridColumn, <System.Runtime.InteropServices.Out()> ByRef rightColumnWhenUnselect As GridColumn)
			rightColumn = Nothing
			rightColumnWhenUnselect = Nothing
			Dim maxWidth As Double = ScrollElement.HorizontalOffset + ScrollElement.ViewportWidth
			Dim columnsWidth As Double = 0R
			rightColumn = FindVisibleColumn(Function(w) w > maxWidth, columnsWidth)
			rightColumnWhenUnselect = FindVisibleColumn(Function(w) w >= maxWidth, columnsWidth)

			If rightColumnWhenUnselect.VisibleIndex - 1 <= 0 Then
				rightColumnWhenUnselect = View.VisibleColumns.First()
			End If

			If rightColumn Is Nothing Then
				rightColumn = View.VisibleColumns.Last()
			End If

			columnsWidth = maxWidth - (columnsWidth - rightColumn.Width)
			If columnsWidth <= rightColumn.Width / 2 Then
				If rightColumn.VisibleIndex - 1 > 0 Then
					rightColumn = View.VisibleColumns(rightColumn.VisibleIndex - 1)
				Else
					rightColumn = View.VisibleColumns.First()
				End If
			End If
		End Sub
		Private Function GetLeftVisibleColumn() As GridColumn
			If ScrollElement.HorizontalOffset < View.VisibleColumns.First().Width / 2 Then
				Return View.VisibleColumns.First()
			End If
			Dim maxWidth As Double = ScrollElement.HorizontalOffset
			Dim columnsWidth As Double = 0R
			Dim res As GridColumn = FindVisibleColumn(Function(w) w > maxWidth, columnsWidth)
			columnsWidth = columnsWidth - res.Width
			If maxWidth < columnsWidth + res.Width / 2 Then
				Return res
			Else
				Return View.VisibleColumns(res.VisibleIndex + 1)
			End If
		End Function
		Private Function FindVisibleColumn(ByVal cond As Func(Of Double, Boolean), ByRef columnsWidth As Double) As GridColumn

			columnsWidth = 0R
			Dim res As GridColumn = Nothing
			If (Not View.AutoWidth) Then
				columnsWidth += LayoutHelper.FindElement(View, AddressOf IsFitContent).ActualWidth
			End If
			For Each gc As GridColumn In View.VisibleColumns
				columnsWidth += gc.ActualHeaderWidth
				If cond(columnsWidth) Then
					res = gc
					Exit For
				End If
			Next gc
			Return res
		End Function
		Private Function IsFitContent(ByVal e As FrameworkElement) As Boolean
			Return e.Name = "PART_FitContent"
		End Function

		Private Function GetRowhandle(ByVal mouseYPosition As Double) As Integer
			Dim avgRowHeight As Double = DataArea.ActualHeight / ScrollElement.ViewportHeight
			Dim currentRowIndex As Integer = 0
			Dim summaryHeight As Double = 0
			Do While summaryHeight < mouseYPosition
				summaryHeight += avgRowHeight
				currentRowIndex += 1
			Loop
			Return Grid.GetRowHandleByVisibleIndex(View.TopRowIndex + currentRowIndex)
		End Function

		Private Sub UpdateSelection()
			Grid.BeginSelection()
			Grid.UnselectAll()
			If Grid.SelectionMode = MultiSelectMode.Row Then
				Grid.SelectRange(StartSelectionInfo.RowHandle, CurrentSelectionInfo.RowHandle)
			Else
				View.SelectCells(StartSelectionInfo.RowHandle, StartSelectionInfo.Column, CurrentSelectionInfo.RowHandle, CurrentSelectionInfo.Column)
			End If
			Grid.EndSelection()
		End Sub

		Private Sub UpdateVerticalScrolling(ByVal pt As Point)
			If Settings.IsMouseOverTopScrollingArea(pt) AndAlso ScrollController.CanScrollUp Then
				ScrollController.StartScrollUp()
			ElseIf Settings.IsMouseOverBottomScrollingArea(pt, DataArea.ActualHeight) AndAlso ScrollController.CanScrollDown Then
				ScrollController.StartScrollDown()
			Else
				ScrollController.StopVerticalScrolling()
			End If
		End Sub
		Private Sub UpdateHorizontalScrolling(ByVal pt As Point)
			If Settings.IsMouseOverLeftScrollingArea(pt) AndAlso ScrollController.CanScrollLeft Then
				ScrollController.StartScrollLeft()
			ElseIf Settings.IsMouseOverRightScrollingArea(pt, DataArea.ActualWidth) AndAlso ScrollController.CanScrollRight Then
				ScrollController.StartScrollRight()
			Else
				ScrollController.StopHorizontalScrolling()
			End If
		End Sub

		Private Sub OnScrollControllerScrollUp(ByVal sender As Object, ByVal e As EventArgs)
			If ScrollController.CanScrollUp Then
				CurrentSelectionInfo.RowHandle = Grid.GetRowHandleByVisibleIndex(View.TopRowIndex + Settings.UnselectedRowCountWhileScrolling)
			Else
				CurrentSelectionInfo.RowHandle = Grid.GetRowHandleByVisibleIndex(View.TopRowIndex)
			End If
			UpdateSelection()
		End Sub
		Private Sub OnScrollControllerScrollDown(ByVal sender As Object, ByVal e As EventArgs)
			If ScrollController.CanScrollDown Then
				CurrentSelectionInfo.RowHandle = Grid.GetRowHandleByVisibleIndex(View.TopRowIndex + ScrollController.VisibleRowCount - Settings.UnselectedRowCountWhileScrolling)
			Else
				CurrentSelectionInfo.RowHandle = Grid.GetRowHandleByVisibleIndex(Grid.VisibleRowCount - 1)
			End If
			UpdateSelection()
		End Sub
		Private Sub OnScrollControllerScrollLeft(ByVal sender As Object, ByVal e As EventArgs)
			If ScrollController.CanScrollLeft Then
				UpdateCurrentSelectionInfo()
				UpdateSelection()
			End If
		End Sub
		Private Sub OnScrollControllerScrollRight(ByVal sender As Object, ByVal e As EventArgs)
			If ScrollController.CanScrollRight Then
				UpdateCurrentSelectionInfo()
			End If
			UpdateSelection()
		End Sub
	End Class
End Namespace

File: CS/DXGridSample/GridSelectingBehavior.cs
using System;
using System.Collections.Generic;
using System.Linq;
using DevExpress.Xpf.Grid;
using System.Windows.Interactivity;
using System.Windows.Input;
using System.Windows;
using System.Windows.Media;
using DevExpress.Xpf.Core.Native;
using System.Windows.Controls.Primitives;

namespace DXGrid_AssignComboBoxToColumn {
    public class SelectionBehaviorSettings {
        public int DragArea = 4;
        public int StartScrollArea = 25;
        public int UnselectedRowCountWhileScrolling = 0;

        public bool IsDragging(Point pt1, Point pt2) {
            return Math.Abs(pt1.X - pt2.X) > DragArea || Math.Abs(pt1.Y - pt2.Y) > DragArea;
        }
        public bool IsMouseOverTopScrollingArea(Point pt) {
            return pt.Y < StartScrollArea;
        }
        public bool IsMouseOverBottomScrollingArea(Point pt, double maxHeight) {
            return pt.Y > maxHeight - StartScrollArea;
        }
        public bool IsMouseOverLeftScrollingArea(Point pt) {
            return pt.X < StartScrollArea;
        }
        public bool IsMouseOverRightScrollingArea(Point pt, double maxWidth) {
            return pt.X > maxWidth - StartScrollArea;
        }
    }
    public class SelectionInfo {
        public Point MousePoint { get; set; }
        public int RowHandle { get; set; }
        public GridColumn Column { get; set; }
        public virtual void Clear() {
            MousePoint = InvalidPoint;
            RowHandle = GridControl.InvalidRowHandle;
            Column = null;
        }
        public virtual bool IsEmptyInfo() {
            return MousePoint == InvalidPoint;
        }

        static readonly Point InvalidPoint = new Point(-10000, 10000);
    }
    public class StartSelectionInfo : SelectionInfo {
        public bool IsSelectionStarted { get; set; }
        public bool IsLeftMouseButtonPressed { get { return Mouse.LeftButton == MouseButtonState.Pressed; } }
        public override void Clear() {
            base.Clear();
            IsSelectionStarted = false;
        }
    }
    public class GridSelectingBehavior : Behavior<GridControl> {
        public SelectionBehaviorSettings Settings { get; set; }
        public GridControl Grid { get { return AssociatedObject; } }
        public TableView View { get { return (TableView)Grid.View; } }
        protected internal IScrollInfo ScrollElement { get; private set; }
        protected internal FrameworkElement DataArea { get; private set; }
        protected StartSelectionInfo StartSelectionInfo { get; private set; }
        protected SelectionInfo CurrentSelectionInfo { get; private set; }
        protected ScrollController ScrollController { get; set; }

        public GridSelectingBehavior() {
            Settings = new SelectionBehaviorSettings();
            StartSelectionInfo = new StartSelectionInfo();
            CurrentSelectionInfo = new SelectionInfo();
            ScrollController = new ScrollController(this);
            ScrollController.ScrollDown += OnScrollControllerScrollDown;
            ScrollController.ScrollUp += OnScrollControllerScrollUp;
            ScrollController.ScrollLeft += OnScrollControllerScrollLeft;
            ScrollController.ScrollRight += OnScrollControllerScrollRight;
        }

        #region  INITIALIZATION
        protected override void OnAttached() {
            base.OnAttached();
            Grid.Loaded += OnGridLoaded;
        }
        protected override void OnDetaching() {
            Grid.PreviewMouseMove -= OnGridPreviewMouseMove;
            Grid.PreviewMouseDown -= OnGridPreviewMouseDown;
            Grid.PreviewMouseUp -= OnGridPreviewMouseUp;
            Grid.Loaded -= OnGridLoaded;
            base.OnDetaching();
        }
        private void OnGridLoaded(object sender, RoutedEventArgs e) {
            View.LayoutUpdated += OnViewLayoutUpdated;
        }
        void OnViewLayoutUpdated(object sender, EventArgs e) {
            if (View.MultiSelectMode == TableViewSelectMode.None)
                throw new Exception("GridSelectingBehavior does not allow using the MultiSelectMode.None.");

            DataArea = LayoutHelper.FindElementByName(View, "PART_ScrollContentPresenter");
            if (DataArea == null) return;
            ScrollElement = (DataPresenter)LayoutHelper.FindElement(DataArea, (el) => el is DataPresenter);
            if (ScrollElement == null) return;

            View.LayoutUpdated -= OnViewLayoutUpdated;

            Grid.PreviewMouseMove += OnGridPreviewMouseMove;
            Grid.PreviewMouseDown += OnGridPreviewMouseDown;
            Grid.PreviewMouseUp += OnGridPreviewMouseUp;
        }
        #endregion

        void OnGridPreviewMouseUp(object sender, MouseButtonEventArgs e) {
            if (!StartSelectionInfo.IsSelectionStarted) return;
            Mouse.Capture(null);
            ScrollController.StopHorizontalScrolling();
            ScrollController.StopVerticalScrolling();
            StartSelectionInfo.Clear();
            CurrentSelectionInfo.Clear();
            e.Handled = true;
        }
        void OnGridPreviewMouseDown(object sender, MouseButtonEventArgs e) {
            StartSelectionInfo.Clear();
            TableViewHitInfo hitInfo = View.CalcHitInfo(e.OriginalSource as DependencyObject);
            if (!hitInfo.InRow && !hitInfo.InRowCell) return;
            StartSelectionInfo.MousePoint = e.GetPosition(DataArea);
            StartSelectionInfo.RowHandle = hitInfo.RowHandle;
            StartSelectionInfo.Column = hitInfo.Column;
        }
        void OnGridPreviewMouseMove(object sender, MouseEventArgs e) {
            if (!StartSelectionInfo.IsLeftMouseButtonPressed || StartSelectionInfo.IsEmptyInfo()) return;
            if (!StartSelectionInfo.IsSelectionStarted) {
                if (!Settings.IsDragging(StartSelectionInfo.MousePoint, e.GetPosition(DataArea))) return;
                Mouse.Capture(DataArea, CaptureMode.SubTree);
                StartSelectionInfo.IsSelectionStarted = true;
            }

            UpdateCurrentSelectionInfo();

            UpdateVerticalScrolling(CurrentSelectionInfo.MousePoint);
            UpdateHorizontalScrolling(CurrentSelectionInfo.MousePoint);

            if (!ScrollController.IsScrollWorking) UpdateSelection();

            ScrollController.UpdateVerticalScrollTimerInterval(DataArea.ActualHeight, e.GetPosition(DataArea).Y);
            ScrollController.UpdateHorizontalScrollTimerInterval(DataArea.ActualWidth, e.GetPosition(DataArea).X);
            e.Handled = true;
        }

        //TODO: improve this method
        void UpdateCurrentSelectionInfo() {
            CurrentSelectionInfo.MousePoint = Mouse.GetPosition(DataArea);
            Point pt = CurrentSelectionInfo.MousePoint;
            //Inside DataArea
            if (pt.X > 0 && pt.Y > 0 && pt.X < DataArea.ActualWidth && pt.Y < DataArea.ActualHeight) {
                HitTestResult result = VisualTreeHelper.HitTest(DataArea, CurrentSelectionInfo.MousePoint);
                if (result != null) {
                    DependencyObject hittedObject = result.VisualHit;
                    TableViewHitInfo hitInfo = View.CalcHitInfo(hittedObject);
                    CurrentSelectionInfo.RowHandle = hitInfo.RowHandle;
                    CurrentSelectionInfo.Column = hitInfo.Column;
                }
                return;
            }
            //Outside DataArea
            CurrentSelectionInfo.RowHandle = GetRowhandle(CurrentSelectionInfo.MousePoint.Y);
            //Select to right
            GridColumn rightColumn, rightColumnWhenUnselect;
            GetRightVisibleColumn(out rightColumn, out rightColumnWhenUnselect);
            if (pt.X >= DataArea.ActualWidth && StartSelectionInfo.Column.VisibleIndex <= CurrentSelectionInfo.Column.VisibleIndex) {
                if (rightColumn.VisibleIndex > CurrentSelectionInfo.Column.VisibleIndex)
                    CurrentSelectionInfo.Column = rightColumn;
                return;
            }
            //Unselecting to right
            if (pt.X >= DataArea.ActualWidth && StartSelectionInfo.Column.VisibleIndex >= CurrentSelectionInfo.Column.VisibleIndex) {
                if (rightColumnWhenUnselect.VisibleIndex >= CurrentSelectionInfo.Column.VisibleIndex)
                    CurrentSelectionInfo.Column = rightColumnWhenUnselect;
                return;
            }
            //Select to left
            if (pt.X <= 0) {
                GridColumn col = GetLeftVisibleColumn();
                    CurrentSelectionInfo.Column = col;
                return;
            }
        }
        void GetRightVisibleColumn(out GridColumn rightColumn, out GridColumn rightColumnWhenUnselect) {
            rightColumn = null;
            rightColumnWhenUnselect = null;
            double maxWidth = ScrollElement.HorizontalOffset + ScrollElement.ViewportWidth;
            double columnsWidth = 0d;
            rightColumn = FindVisibleColumn((w) => w > maxWidth, ref columnsWidth);
            rightColumnWhenUnselect = FindVisibleColumn((w) => w >= maxWidth, ref columnsWidth);

            if (rightColumnWhenUnselect.VisibleIndex - 1 <= 0)
                rightColumnWhenUnselect = View.VisibleColumns.First();

            if (rightColumn == null) rightColumn = View.VisibleColumns.Last();

            columnsWidth = maxWidth - (columnsWidth - rightColumn.Width);
            if (columnsWidth <= rightColumn.Width / 2) {
                if (rightColumn.VisibleIndex - 1 > 0)
                    rightColumn = View.VisibleColumns[rightColumn.VisibleIndex - 1];
                else rightColumn = View.VisibleColumns.First();
            }
        }
        GridColumn GetLeftVisibleColumn() {
            if (ScrollElement.HorizontalOffset < View.VisibleColumns.First().Width / 2) return View.VisibleColumns.First();
            double maxWidth = ScrollElement.HorizontalOffset;
            double columnsWidth = 0d;
            GridColumn res = FindVisibleColumn((w) => w > maxWidth, ref columnsWidth);
            columnsWidth = columnsWidth - res.Width;
            if (maxWidth < columnsWidth + res.Width / 2)
                return res;
            else
                return View.VisibleColumns[res.VisibleIndex + 1];
        }
        GridColumn FindVisibleColumn(Func<double, bool> cond, ref double columnsWidth) {
            columnsWidth = 0d;
            GridColumn res = null;
            foreach (GridColumn gc in View.VisibleColumns) {
                columnsWidth += gc.Width;
                if (cond(columnsWidth)) {
                    res = gc;
                    break;
                }
            }
            return res;
        }


        int GetRowhandle(double mouseYPosition) {
            double avgRowHeight = DataArea.ActualHeight / ScrollElement.ViewportHeight;
            int currentRowIndex = 0;
            double summaryHeight = 0;
            while (summaryHeight < mouseYPosition) {
                summaryHeight += avgRowHeight;
                currentRowIndex++;
            }
            return Grid.GetRowHandleByVisibleIndex(View.TopRowIndex + currentRowIndex);
        }

        void UpdateSelection() {
            View.BeginSelection();
            View.ClearSelection();
            if (View.MultiSelectMode == TableViewSelectMode.Row)
                View.SelectRange(StartSelectionInfo.RowHandle, CurrentSelectionInfo.RowHandle);
            else
                View.SelectCells(StartSelectionInfo.RowHandle, StartSelectionInfo.Column, CurrentSelectionInfo.RowHandle, CurrentSelectionInfo.Column);
            View.EndSelection();
        }

        void UpdateVerticalScrolling(Point pt) {
            if (Settings.IsMouseOverTopScrollingArea(pt) && ScrollController.CanScrollUp)
                ScrollController.StartScrollUp();
            else if (Settings.IsMouseOverBottomScrollingArea(pt, DataArea.ActualHeight) && ScrollController.CanScrollDown)
                ScrollController.StartScrollDown();
            else
                ScrollController.StopVerticalScrolling();
        }
        void UpdateHorizontalScrolling(Point pt) {
            if (Settings.IsMouseOverLeftScrollingArea(pt) && ScrollController.CanScrollLeft)
                ScrollController.StartScrollLeft();
            else if (Settings.IsMouseOverRightScrollingArea(pt, DataArea.ActualWidth) && ScrollController.CanScrollRight)
                ScrollController.StartScrollRight();
            else ScrollController.StopHorizontalScrolling();
        }
        
        void OnScrollControllerScrollUp(object sender, EventArgs e) {
            if (ScrollController.CanScrollUp)
                CurrentSelectionInfo.RowHandle = Grid.GetRowHandleByVisibleIndex(View.TopRowIndex + Settings.UnselectedRowCountWhileScrolling);
            else CurrentSelectionInfo.RowHandle = Grid.GetRowHandleByVisibleIndex(View.TopRowIndex);
            UpdateSelection();
        }
        void OnScrollControllerScrollDown(object sender, EventArgs e) {
            if (ScrollController.CanScrollDown)
                CurrentSelectionInfo.RowHandle = Grid.GetRowHandleByVisibleIndex(View.TopRowIndex + ScrollController.VisibleRowCount - Settings.UnselectedRowCountWhileScrolling);
            else CurrentSelectionInfo.RowHandle = Grid.GetRowHandleByVisibleIndex(Grid.VisibleRowCount - 1);
            UpdateSelection();
        }
        void OnScrollControllerScrollLeft(object sender, EventArgs e) {
            if (ScrollController.CanScrollLeft) {
                UpdateCurrentSelectionInfo();
                UpdateSelection();
            }
        }
        void OnScrollControllerScrollRight(object sender, EventArgs e) {
            if (ScrollController.CanScrollRight)
                UpdateCurrentSelectionInfo();
                UpdateSelection();
        }
    }

}


